packages/medusa/src/services/event-bus.ts

```diff
@@ -167,33 +209,48 @@ export default class EventBusService {
   async emit<T>(
     eventName: string,
     data: T,
-    options: EmitOptions = {}
+    options: EmitOptions = { attempts: 1 }
   ): Promise<StagedJob | void> {
+    const opts: { removeOnComplete: boolean } & EmitOptions = {
+      removeOnComplete: true,
+      attempts: 1,
+    }
+    if (typeof options.attempts === "number") {
+      opts.attempts = options.attempts
+      if (typeof options.backoff !== "undefined") {
+        opts.backoff = options.backoff
+      }
+    }
+    if (typeof options.delay === "number") {
+      opts.delay = options.delay
+    }
+
+    /**
+     * If we are in an ongoing transaction, we store the jobs in the database
+     * instead of processing them immediately. We only want to process those
+     * events, if the transaction successfully commits. This is to avoid jobs
+     * being processed if the transaction fails.
+     *
+     * In case of a failing transaction, kobs stored in the database are removed
+     * as part of the rollback.
+     */
     if (this.transactionManager_) {
       const stagedJobRepository = this.transactionManager_.getCustomRepository(
         this.stagedJobRepository_
       )
 
-      const stagedJobInstance = stagedJobRepository.create({
+      const jobToCreate: Partial<StagedJob> = {
         event_name: eventName,
-        data,
-      } as StagedJob)
-      return await stagedJobRepository.save(stagedJobInstance)
-    } else {
-      const opts: { removeOnComplete: boolean } & EmitOptions = {
-        removeOnComplete: true,
-      }
-      if (typeof options.attempts === "number") {
-        opts.attempts = options.attempts
-        if (typeof options.backoff !== "undefined") {
-          opts.backoff = options.backoff
-        }
-      }
-      if (typeof options.delay === "number") {
-        opts.delay = options.delay
+        data: data as Record<string, unknown>,
```

Comment:
suggsetion(non blocking): should we change the type of the data in the argument to be Record<string, unknown> since we are casting it here? From my pov it means that it can't be something else than that. wdyt?